pkgs/by-name/wo/wordlists/package.nix

```diff
@@ -0,0 +1,70 @@
+{ lib
+, callPackage
+, nmap
+, rockyou
+, runtimeShell
+, seclists
+, symlinkJoin
+, tree
+, wfuzz
+, lists ? [
+    nmap
+    rockyou
+    seclists
+    wfuzz
+  ]
+}:
+
+symlinkJoin rec {
```

Comment:
I wrote this in a response in one of the outdated threads, but I think it regards more than the initial discussion, so I thought I'd rather make a new thread.

I've had another look at this. Introducing binaries in the middle of a `symlinkJoin` feels like a hack that circumvents the purpose of `symlinkJoin`.

I think the canonical way would've been to let the original binaries be standalone tools that takes their input via normal methods like args, envvars or stdin, and then rather wrap the program to fix its inputs. Or at least somehow build the tools in standalone derivations that depend on the lists as input.

How about a design that looks more like this?

```mermaid
flowchart TD
    wordlists["wordlists package (symlinkJoin)"]
    binaries["binaries (wrapping wordlist collection dir)"]
    symlinkedWordlists["wordlist collection (symlinkJoin)"]
    wordlist1["wordlist 1"]
    wordlist2["wordlist 2"]
    wordlist3["wordlist 3"]

    wordlists --> binaries
    wordlists --> symlinkedWordlists
    binaries --> symlinkedWordlists
    symlinkedWordlists --> wordlist1
    symlinkedWordlists --> wordlist2
    symlinkedWordlists --> wordlist3
```

This would resolve several of the comments that are currently "unfixable" due to the weird use of `symlinkJoin`

Reply:
> So wordlists package (symlinkJoin) would be a package including wordlists or would that be the current wordlists package we have?

This would be the equivalent of the current wordlists package we already have, consisting of the `symlinkJoin` of the binaries derivation and the "wordlist collection" derivation

> binaries (wrapping wordlist collection dir) this would be binaries interacting with wordlists like the current wordlists and wordlists_path ones?

Correct. And this derivation will use the "wordlist collection" as its input, so it can refer to it without needing to refer to the `$out` of its own derivation.

> What is wordlist collection (symlinkJoin) supposed to be?

This is a derivation created with `symlinkJoin` which only contains the combination of the wordlists, no binaries and nothing else. The binaries derivation will be able to refer to this derivation as its input.

Reply:
>> What is wordlist collection (symlinkJoin) supposed to be?
> This is a derivation created with symlinkJoin which only contains the combination of the wordlists, no binaries and nothing else. The binaries derivation will be able to refer to this derivation as its input.

what would you do about packages like nmap that themselfs have a binary as a output? would you go and do a extra derivation that only includes the wordlist or would you be fine with the way we are currently doing it?

Reply:
Thanks for elaborating, I'll start working on a implantation.

Reply:
> what would you do about packages like nmap that themselfs have a binary as a output? would you go and do a extra derivation that only includes the wordlist or would you be fine with the way we are currently doing it?

For cleanliness sake, I think I'd create the derivation not using `symlinkJoin`, but rather using either `runCommand` or `stdenvNoCC.mkDerivation` and then symlinking the contents inside every packages `$out/share/wordlists`.

However, using `symlinkJoin` shouldn't break anything as long as there's no colliding paths AFAIK.

Reply:
Hmm okay, this feels like it's adding more complexity then it's worth because with the way you suggested if we still want to allow the user to override the lists used we have to give the list of wordlists provided in `wordlists package` to every single `wordlists collection` which seems like a bit of ugly wiring and we also have to provide every single `binaries (wrapping wordlist collection dir)` with the current collection. which this both wouldn't be hard to do it's just adding a lot of boilerplate and seems to over complicate stuff. I don't really see the problem this would fix/the value we would get.

Reply:
I'm a bit confused.

> every single wordlists collection

The wordlists collection is a single derivation consisting of the list of wordlists the user provides. wordlist1, wordlist2, and wordlist3 are just example substitutes for these. The suggestion about cleaning the list of wordlist derivations is separate from the architectural change, the result would be the same with and without.

The only real change here is that we split up the large `symlinkJoin` step into 3 pieces. One for joining the wordlists, one for creating the binaries, and lastly one for the combining these two. We don't need to change how the user should be able to override the list of wordlists.

This is the corresponding diagram for the earlier implementation for reference:

```mermaid
flowchart TD
    wordlists["wordlists package (symlinkJoin)\ncontains binaries injected in postBuild"]
    wordlist1["wordlist 1"]
    wordlist2["wordlist 2"]
    wordlist3["wordlist 3"]

    wordlists --> wordlist1
    wordlists --> wordlist2
    wordlists --> wordlist3
```

Reply:
I think the current code is good enough and we can refactor/make it better at a later point. The binaries could be extracted to a let statement, but that would just be internal refactoring in the derivation. There would be no point in putting the scripts into the global package namespace anyway since their usecase would be very limited.